service/hproxy/hproxy.go

```diff
@@ -0,0 +1,960 @@
+
+		j := make(map[string]any)
+		err = json.Unmarshal(resp, &j)
+		if err != nil {
+			return err
+		}
+
+		// Make sure we have "result"
+		if _, ok := j["result"]; !ok {
+			return errors.New("no result")
+		}
+		return nil
+	})
+
+	s.hvmHandlers = append(s.hvmHandlers, hvmHandler)
+
+	return nil
+}
+
+func (s *Server) nodeRemove(node string) error {
+	u, err := url.Parse(node)
+	if err != nil {
+		return fmt.Errorf("invalid url: %w", err)
+	}
+	switch u.Scheme {
+	case "http", "https":
+	default:
+		return fmt.Errorf("unsuported scheme: %v", u.Scheme)
+	}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	// make sure it isn't a dupe
+	for k := range s.hvmHandlers {
+		if s.hvmHandlers[k].u.String() == u.String() {
+			if s.hvmHandlers[k].state != StateRemoved {
+				s._nodeRemoved(k)
+				return nil
+			}
+			return errors.New("already removed")
+		}
+	}
+	return errors.New("not found")
+}
+
+// _countNode must be called with the mutext held.
+func (s *Server) _nodeCount(id int) int {
+	var count int
+	for _, v := range s.clients {
+		if v.node == id {
+			count++
+		}
+	}
+	return count
+}
+
+// _nodeReap must be called with the mutext held.
+func (s *Server) _nodeReap(id int) {
+	for k, v := range s.clients {
+		if v.node == id {
+			delete(s.clients, k)
+		}
+	}
+}
+
+// _nodeNew must be called with lock held.
+func (s *Server) _nodeNew(id int) {
+	if s.hvmHandlers[id].state != StateNew {
+		s.hvmHandlers[id].connections = 0 // reset connections
+		s.hvmHandlers[id].state = StateNew
+		s._nodeReap(id)
+		log.Infof("Marking hvm new %v: %v", id, s.hvmHandlers[id].u)
+	}
+}
+
+// _nodeRemoved must be called with lock held.
+func (s *Server) _nodeRemoved(id int) {
+	if s.hvmHandlers[id].state != StateRemoved {
+		s.hvmHandlers[id].connections = 0 // reset connections
+		s.hvmHandlers[id].state = StateRemoved
+		s._nodeReap(id)
+		log.Infof("Marking hvm removed %v: %v", id, s.hvmHandlers[id].u)
+	}
+}
+
+func (s *Server) nodeHealthy(id int) {
+	s.mtx.Lock()
+	if s.hvmHandlers[id].state != StateHealthy {
+		s.hvmHandlers[id].connections = 0 // reset connections
+		s.hvmHandlers[id].state = StateHealthy
+		s._nodeReap(id)
+		log.Infof("Marking hvm healthy %v: %v", id, s.hvmHandlers[id].u)
+	}
+	s.mtx.Unlock()
+}
+
+func (s *Server) nodeUnhealthy(id int) {
+	s.mtx.Lock()
+	if s.hvmHandlers[id].state != StateUnhealthy {
+		s.hvmHandlers[id].connections = 0 // reset connections
+		s.hvmHandlers[id].state = StateUnhealthy
+		s._nodeReap(id)
+		log.Infof("Marking hvm unhealthy %v: %v", id, s.hvmHandlers[id].u)
+	}
+	s.mtx.Unlock()
+}
+
+func (s *Server) poke(ctx context.Context, id int) {
+	log.Tracef("poke: %v", id)
+	defer log.Tracef("poke exit: %v", id)
+
+	s.mtx.Lock()
+	if s.hvmHandlers[id].poking {
+		s.mtx.Unlock()
+		return
+	}
+	s.hvmHandlers[id].poking = true
+	poker := s.hvmHandlers[id].poker
+	s.mtx.Unlock()
+	defer func() {
+		s.mtx.Lock()
+		s.hvmHandlers[id].poking = false
+		s.mtx.Unlock()
+	}()
+
+	err := poker.Poke(ctx)
+	if err != nil {
```

Comment:
let's add a trace log here to debug why nodes are being marked "unhealthy", when needed